File: answers/leetcode/Implement Trie (Prefix Tree).cpp
//@type Trie Design
//@result 14 / 14 test cases passed. Status: Accepted Runtime: 60 ms Submitted: 2 minutes ago You are here! Your runtime beats 62.25% of cpp submissions.

class TrieNode {
public:
    // Initialize your data structure here.
    TrieNode() : key_flag(false) {
        for (int i = 0; i < 26; ++ i) {
			children[i] = NULL;
		}
    }
	bool key_flag;
	TrieNode *children[26];
};

class Trie {
public:
    Trie() {
        root = new TrieNode();
    }

    // Inserts a word into the trie.
    void insert(string word) {
        TrieNode *node = root;
		for (size_t i = 0; i < word.length(); ++ i) {
			if (! node->children[word[i] - 'a']) {
				node->children[word[i] - 'a'] = new TrieNode();
			}
			node = node->children[word[i] - 'a'];
		}
		node->key_flag = true;
    }

    // Returns if the word is in the trie.
    bool search(string word) {
        TrieNode *node = root;
		for (size_t i = 0; i < word.length(); ++ i) {
			if (node) {
				node = node->children[word[i] - 'a'];
			}
			else {
				break;
			}
		}
		return node ? node->key_flag : false;
    }

    // Returns if there is any word in the trie
    // that starts with the given prefix.
    bool startsWith(string prefix) {
        TrieNode *node = root;
		for (size_t i = 0; i < prefix.length(); ++ i) {
			if (node) {
				node = node->children[prefix[i] - 'a'];
			}
			else {
				break;
			}
		}
		return node;
    }

private:
    TrieNode* root;
};

// Your Trie object will be instantiated and called as such:
// Trie trie;
// trie.insert("somestring");
// trie.search("key");

File: answers/leetcode/Implement Trie (Prefix Tree).py
class TrieNode(object):
    def __init__(self):
        """
        Initialize your data structure here.
        """

        self.char_map = {}
        self.is_word = False

class Trie(object):

    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        """
        Inserts a word into the trie.
        :type word: str
        :rtype: void
        """

        current = self.root
        for i in range(len(word)):
            if word[i] not in current.char_map:
                current.char_map[word[i]] = TrieNode()
            current = current.char_map[word[i]]
        current.is_word = True

    def search(self, word):
        """
        Returns if the word is in the trie.
        :type word: str
        :rtype: bool
        """

        current = self.root
        for i in range(len(word)):
            if word[i] not in current.char_map:
                return False
            current = current.char_map[word[i]]
        return current.is_word

    def startsWith(self, prefix):
        """
        Returns if there is any word in the trie
        that starts with the given prefix.
        :type prefix: str
        :rtype: bool
        """

        current = self.root
        for i in range(len(prefix)):
            if prefix[i] not in current.char_map:
                return False
            current = current.char_map[prefix[i]]
        return True

# Your Trie object will be instantiated and called as such:
# trie = Trie()
# trie.insert("somestring")
# trie.search("key")
